syft/rekor/query.go

```diff
@@ -0,0 +1,182 @@
+package rekor
+
+import (
+	"context"
+	"crypto"
+	"crypto/sha1"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/file"
+	"github.com/anchore/syft/syft/source"
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/entries"
+	"github.com/sigstore/rekor/pkg/generated/client/index"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/spdx/tools-golang/spdx"
+)
+
+// getSbom attempts to retrieve the SBOM from the URI.
+//
+// Precondition: att is not nil and client is not nil
+func getSbom(att *InTotoAttestation, client *http.Client) (*[]byte, error) {
+	if len(att.Predicate.Sboms) > 1 {
+		log.Info("Attestation found on Rekor with multiple SBOMS, which is not currently supported. Proceeding with the first SBOM.")
+	}
+	uri := att.Predicate.Sboms[0].Uri
+
+	resp, err := client.Get(uri)
+	if err != nil {
+		return nil, fmt.Errorf("error making http request: %w", err)
+	}
+
+	if resp == nil {
+		return nil, errors.New("http response body is nil")
+	}
+	bytes, err := io.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("error reading http response: %w", err)
+	}
+
+	return &bytes, nil
+}
+
+// getUuids returns the uuids of the Rekor entries associated with an sha hash
+//
+// Precondition: client is not nil
+func getUuids(sha string, client *client.Rekor) ([]string, error) {
+	query := &models.SearchIndex{Hash: sha}
+	var params *index.SearchIndexParams = index.NewSearchIndexParams().WithQuery(query)
+
+	res, err := client.Index.SearchIndex(params)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, errors.New("result of Rekor search is nil")
+	}
+	return res.Payload, nil
+}
+
+// Precondition: Client is not nil
+func getRekorEntry(uuid string, client *client.Rekor) (models.LogEntry, error) {
+	params := entries.NewGetLogEntryByUUIDParams().WithEntryUUID(uuid)
+	res, err := client.Entries.GetLogEntryByUUID(params)
+	if err != nil {
+		return nil, err
+	}
+	if res == nil {
+		return nil, errors.New("result of Rekor query is nil")
```

Comment:
out of curiousity do you want it to be an error when there are no rekor entries associated to an sbom, or no? (here and above)

Reply:
I consider `res` being `nil` here an error because we only query rekor for uuids that we know exist. Similarly above in `getUuids` also an error because the expected behavior is for res.Payload to be nil, not res. 

When there are rekor entries returned but none are associated with SBOMs, that is handled here https://github.com/mdeicas/syft/blob/query-rekor/syft/rekor/utils.go#L125. In this case I treat this as an error so the  message can be logged later on. 

Let me know if you have any suggestions on how I handle errors though. Its currently designed with informative logging in mind, but it doesn't really do a good job of separating actual, unexpected errors from more expected behavior such as not finding any rekor entries.